feat: datasource exoscale

mcorbin opened the pull request:
Create a new datasource for Exoscale.

I took inspiration from the Cloudstack datasource, did some cleanup, and used the `169.254.169.254` address to fetch the metadata/userdata/password.

mcorbin committed:
feat: datasource exoscale

chrisglass commented on tools/ds-identify:
```diff
@@ -553,6 +553,12 @@ dscheck_CloudStack() {
 
+dscheck_Exoscale() {
+    is_container && return ${DS_NOT_FOUND}
+    dmi_product_name_matches "Exoscale*" && return $DS_FOUND
```
Note: this will not work right now (we have a cloudstack product name), but that won't be a problem since we instead will rely on the datasource being set in the templates, in cloud.cfg.d/

We will change our DMI product name accordingly, and so it will work in the future in the general case (once a newer cloud-init trickles down to all the templates).

chrisglass replied:
This is fine in this branch - however we must ensure the DMI product name is set to "Exoscale*" before this lands in upstream cloud-init.

This also means we need to document this behavior properly when opening up something like "register custom templates to Exoscale" to the general public: template uploaders must either ensure that their version of cloud-init is recent enough, or they must supply the datasource as part of cloud.cfg.d/ inside their template.

Commit:
cc_rsyslog: Escape possible nested set

Under Python 3.7, we are seeing `FutureWarning: Possible nested set at
position 23`; escaping this bracket causes that warning to disappear.

LP: #1816967

Commit:
util: don't determine string_types ourselves

six already provides this for us, and we're already paying the cost to
determine it there; no need to do it twice.

Commit:
tests: fix some slow tests and some leaking state

In test_ds_identify, don't mutate otherwise-static test data. When
running tests in a random order, this was causing failures due to
breaking preconditions for other tests.

In tests/helpers, reset logging level in tearDown. Some of the CLI
tests set the level of the root logger in a way that isn't correctly
reset.

For test_poll_imds_re_dhcp_on_timeout and
test_dhcp_discovery_run_in_sandbox_warns_invalid_pid, mock out
time.sleep; this saves ~11 seconds (or ~40% of previous test time!).

Commit:
cc_apt_pipelining: stop disabling pipelining by default

This was introduced due to Ubuntu using S3 mirrors, and S3 having a
buggy pipelining implementation.  Those Ubuntu mirrors are no longer in
production and, furthremore, apt has also grown the ability to handle
servers with broken pipelining.

As such, we can stop disabling pipelining, which should result in
improved apt download speeds.

LP: #1794982

chrisglass commented on cloudinit/sources/DataSourceExoscale.py:
```diff
@@ -0,0 +1,125 @@
+
+    def __init__(self, sys_cfg, distro, paths):
+        sources.DataSource.__init__(self, sys_cfg, distro, paths)
+        LOG.info("initialize the Exoscale datasource")
```
```suggestion
        LOG.info("Initializing the Exoscale datasource")
```

chrisglass commented on cloudinit/sources/DataSourceExoscale.py:
```diff
@@ -0,0 +1,125 @@
+        self.extra_config = {}
+
+    def get_password(self):
+        """return the VM password"""
```
```suggestion
        """Return the VM's password."""
```

chrisglass commented on cloudinit/sources/DataSourceExoscale.py:
```diff
@@ -0,0 +1,125 @@
+
+    def get_password(self):
+        """return the VM password"""
+        LOG.info("fetching password")
```
```suggestion
        LOG.info("Fetching password from metadata service")
```

chrisglass commented on cloudinit/sources/DataSourceExoscale.py:
```diff
@@ -0,0 +1,125 @@
+        if password in ['', 'saved_password']:
+            LOG.info("password is missing or already saved")
+            return None
+        LOG.info("found the password, save it")
```
```suggestion
        LOG.info("Found the password in metadata service")
```

chrisglass commented on cloudinit/sources/DataSourceExoscale.py:
```diff
@@ -0,0 +1,125 @@
+        return password
+
+    def wait_for_metadata_service(self):
+        """wait for the metadata service"""
```
```suggestion
        """Wait for the metadata service to be reachable."""
```

chrisglass commented on cloudinit/sources/DataSourceExoscale.py:
```diff
@@ -0,0 +1,125 @@
+            return False
+
+    def _get_data(self):
+        """fetch the user data, the metadata and the VM password"""
```
```suggestion
        """Fetch the user data, the metadata and the VM password from the metadata service."""
```

chrisglass commented on tests/unittests/test_datasource/test_exoscale.py:
```diff
@@ -0,0 +1,85 @@
+        self.userdata_url = "{}/{}/user-data".format(SERVICE_ADDRESS,
+                                                     API_VERSION)
+
+    def test_password_saved(self):
```
```suggestion
    def test_password_saved(self):
        """The password is not set when it is not found in the metadata service."""
```

chrisglass commented on tests/unittests/test_datasource/test_exoscale.py:
```diff
@@ -0,0 +1,85 @@
+                               body="saved_password")
+        self.assertFalse(ds.get_password())
+
+    def test_password_empty(self):
```
```suggestion
    def test_password_empty(self):
        """No password is set if the metadata service returns an empty string."""
```

chrisglass commented on tests/unittests/test_datasource/test_exoscale.py:
```diff
@@ -0,0 +1,85 @@
+                               body="")
+        self.assertFalse(ds.get_password())
+
+    def test_password(self):
```
```suggestion
    def test_password(self):
        """The password is set to what is found in the metadata service."""
```

chrisglass commented on tests/unittests/test_datasource/test_exoscale.py:
```diff
@@ -0,0 +1,85 @@
+        password = ds.get_password()
+        self.assertEqual(expected_password, password)
+
+    def test_get_data(self):
```
```suggestion
    def test_get_data(self):
        """The datasource conforms to expected behavior when supplied full test data."""
```

chrisglass reviewed (approved):
This looks good! +1

I added a lot of pedantic comments (sorry) about capitalization/[PEP257](https://www.python.org/dev/peps/pep-0257/#one-line-docstrings), but those are not blocking.

mcorbin committed:
fixup! feat: datasource exoscale

mcorbin commented:
Thanks for the review @chrisglass ;)

kstieger committed:
net: append type:dhcp[46] only if dhcp[46] is True in v2 netconfig

When providing netplan configuration to cloud-init, the internal
network state would enable DHCP if the 'dhcp' key was present in
the source config.  In netplan, dhcp[46] is a boolean and the
value of the boolean should control whether DHCP is enabled rather
than the presence of the key.  This issue leaded to  inconsistant
sysconfig/network-scripts on fedora. 'BOOTPROTO' was always 'dhcp',
even if the address config was static.

After this change a dhcp subnet is added only if the 'dhcp' setting
in source cfg dict is True.

LP: #1818032

raharper committed:
tip-pylint: Fix assignment-from-return-none errors

pylint now complains about assignment of None from a
return of a function call.  This does not account for subclassing
so we resolve this issue by removing the assignment in the unittest.

Commit:
drop Python 2.6 support and our NIH version detection

- Remove the last few places that use `if PY26`
- Replace our Python version detection logic with six's (which we were
  already using in most places)

Commit:
helpers/openstack: Treat unknown link types as physical

Some deployments of OpenStack expose link types to the guest which
cloud-init doesn't recognise. These will almost always be physical, so
we can operate more robustly if we assume that they are (whilst warning
the user that we're seeing something unexpected).

LP: #1639263

Commit:
clean: correctly determine the path for excluding seed directory

Previously, init.paths.cloud_dir has a trailing slash, which meant that
"/var/lib/cloud//seed" was being compared to "/var/lib/cloud/seed" and
(of course), never matching.

In this commit, switch to using os.path.join to avoid this case (and
update the tests to catch it in future).

LP: #1818571

reddare committed:
Example for Microsoft Azure data disk added.

smoser committed:
Support locking user with usermod if passwd is not available.

In some cases, the 'passwd' command might not be available, but
'usermod' might be.  In debian systems both are provided by the
'passwd' package.  In Redhat/Centos passwd comes from 'passwd' package
while 'usermod' comes from `shadow-utils`

This should just support either one with no real cost other than
the check.

raharper committed:
cloud-init-per: POSIX sh does not support string subst, use sed

cloud-init-per is run via /bin/sh which requires POSIX shell
compliance and does not implement string substitution like
bash.  Replace these calls with use of sed.

LP: #1819222

Commit:
DataSourceEc2: update RELEASE_BLOCKER to be more accurate

Our previous understanding of the upgrade issue was incomplete; it turns
out the only change we need is the one now outlined.

rjschwei committed:
net/sysconfig: Handle default route setup for dhcp configured NICs

When the network configuration has a default route configured and
another network device that is configured with dhcp, SUSE sysconfig
output should not accept the default route provided by the dhcp
server.

LP: #1812117

Commit:
DataSource: move update_events from a class to an instance attribute

Currently, DataSourceAzure updates self.update_events in __init__. As
update_events is a class attribute on DataSource, this updates it for
all instances of classes derived from DataSource including those for
other clouds. This means that if DataSourceAzure is even instantiated,
its behaviour is applied to whichever data source ends up being used for
boot.

To address this, update_events is moved from a class attribute to an
instance attribute (that is therefore populated at instantiation time).
This retains the defaults for all DataSource sub-class instances, but
avoids them being able to mutate the state in instances of other
DataSource sub-classes.

update_events is only ever referenced on an instance of DataSource (or a
sub-class); no code relies on it being a class attribute. (In fact,
it's only used within methods on DataSource or its sub-classes, so it
doesn't even _need_ to remain public, though I think it's appropriate
for it to be public.)

DataSourceScaleway is also updated to move update_events from a
class attribute to an instance attribute, as the class attribute would
now be masked by the DataSource instance attribute.

LP: #1819913

Commit:
tox: bump pylint version to latest (2.3.1)

The previous version was emitting errors due to an incompatibility with
one of its dependencies.

(We could have pinned the dependency instead, but staying current on
pylint is a worthy goal in and of itself.)

Commit:
doc: Refresh Azure walinuxagent docs

- Remove outdated waagent.conf recommendations
- Recommend using Provisioning.UseCloudInit
- Reorganise sections so walinuxagent recommendations are easier to find

Commit:
Add ubuntu_drivers config module

The ubuntu_drivers config module enables usage of the 'ubuntu-drivers'
command. At this point it only serves as a way of installing NVIDIA
drivers for general purpose graphics processing unit (GPGPU)
functionality.

Also, a small usability improvement to get_cfg_by_path to allow it to
take a string for the key path
  "toplevel/second/mykey"
in addition to the original:
  ("toplevel", "second", "mykey")

Commit:
net: Fix ipv6 static routes when using eni renderer

When rendering ipv6 static routes in eni format the
post-up/pre down commands were not correct for ipv6.

LP: #1818669

toabctl committed:
tox: Update testenv for openSUSE Leap to 15.0

Use the requirements for the openSUSE Leap 15.0 release.

rjschwei committed:
net/sysconfig: write out SUSE-compatible IPv6 config

For writing IPv6 addresses to ifcfg-* the name "IPV6ADDR" is used. For
secondary IPs the value for "IPV6ADDR_SECONDARIES" is set. On SUSE based
distributions the names "IPADDR6" and "IPADDR6_$SOMELABEL" need to be
used.

jasonzio committed:
Azure: Ensure platform random_seed is always serializable as JSON.

The Azure platform surfaces random bytes into /sys via Hyper-V.
Python 2.7 json.dump() raises an exception if asked to convert
a str with non-character content, and python 3.0 json.dump()
won't serialize a "bytes" value. As a result, c-i instance
data is often not written by Azure, making reboots slower (c-i
has to repeat work).

The random data is base64-encoded and then decoded into a string
(str or unicode depending on the version of Python in use). The
base64 string has just as many bits of entropy, so we're not
throwing away useful "information", but we can be certain
json.dump() will correctly serialize the bits.

raharper committed:
cloud_tests: fix apt_pipelining test-cases

The apt_pipelining test-cases were broken but until cloud-init
changed it's default behavior to not disable, these silently passed
as both only ever checked if pipelinging was disabled.

First, the tests used the 'apt' namespace, which is not for
configuring pipelining, rather that requires 'apt_pipelining'
 as the namespace.

Second, the 'os' variant needs to check that cloud-init does not
write a configuration file; it was a copy-and-paste error from the
disable test-case.

This branch fixes the config and collection to validate both
scenarios.

anhvoms committed:
DatasourceAzure: add additional logging for azure datasource

Create an Azure logging decorator and use additional ReportEventStack
context managers to provide additional logging details.

jasonzio committed:
Azure: Treat _unset network configuration as if it were absent

When the Azure datasource persists all of its metadata to the
instance directory, it deliberately sets the self.network_config
value to be the sources.UNSET value. The goal is to ensure that
each time the system boots, fresh network configuration data is
fetched from the cloud platform so that any control plane changes
will take effect. When a VM is first created, there's no pickled
instance to restore, so self._network_config is None, resulting
in self.network_config() properly building a new config. Azure
suffered from LP: #1801364 which prevented ds from being stored
in obj.pkl in the instance directory, so subsequent reboots always
regenerated their network configuration.

Commit 0dc3a77f41f4544e4cb5a41637af7693410d4cdf introduced a
new bug in which self.network_config() assumed the
self._network_config value was either None or trustable; when
the config was unpickled, that value was _unset, thus breaking
the assumption.

LP: #1823084

blackboxsw committed:
ubuntu_advantage: rewrite cloud-config module

ubuntu-advantage-tools version 19 has a different command line
interface. Update cloud-init's config module to accept new
ubuntu_advantage configuration settings.

* Underscores better than hyphens: deprecate 'ubuntu-advantage'
  cloud-config key in favor of 'ubuntu_advantage'
* Attach machines with either sso credentials of UA user_token
* Services are enabled by name though an 'enable' list
* Raise warnings if deprecated ubuntu-advantage config keys are
  present, or errors if its config we cannott adapt to

Ubuntu Advantage support can now be configured via #cloud-config
with the following yaml:

ubuntu_advantage:
  token: 'thisismyubuntuadvantagetoken'
  enable: [esm, fips, livepatch]

Co-Authored-By: Daniel Watkins <daniel.watkins@canonical.com>

Commit:
cmd:main.py: Fix missing 'modules-init' key in modes dict

Cloud-init's main.py will fail when presented with a new
stage name 'modules-init' if upgrading an older cloud-init.
Fix this by initializing unknown stage names before accessing.

LP: #1815109

zharalim committed:
Change DataSourceNoCloud to ignore file system label's case.

NoCloud data source now accepts both 'cidata' and 'CIDATA'
as filesystem labels. This is similar to DataSourceConfigDrive's
support for 'config-2' and 'CONFIG-2'.

Commit:
Revert "DataSource: move update_events from a class to an instance..."

Moving update_events from a class attribute to an instance attribute
means that it doesn't exist on DataSource objects that are unpickled,
causing tracebacks on cloud-init upgrade.

As this change is only required for cloud-init installations which don't
utilise ds-identify, we're backing it out to be reintroduced once the
upgrade path bug has been addressed.

This reverts commit f2fd6eac4407e60d0e98826ab03847dda4cde138.

do3meli committed:
cc_apt_configure: fix typo in apt documentation

goneri committed:
mount_cb: do not pass sync and rw options to mount

On FreeBSD, mount_cd9660 does not accept the sync option that is enabled
by default. In addition, the sync is only useful with the `rw` mode.
However the `rw` mode was never used.

This patch removes the `rw` and `sync` parameter of `mount_cb` to
simplify the code base and resolve the FreeBSD issue.

LP: #1645824

Commit:
setup.py: install bash completion script in new location

Per lintian, this is the path at which bash completion scripts should
now be installed.

jasonzio committed:
test_azure: mock util.SeLinuxGuard where needed

Mock util.SeLinuxGuard to do nothing within tests that mock functions
used by the guard, when those mocks confuse the guard. This has no
impact when executing unit tests on systems which do not enable selinux
(e.g. Ubuntu).

LP: #1825253

Commit:
packages: update rpm specs for new bash completion path

LP: #1825444

raharper committed:
net/sysconfig: only indicate available on known sysconfig distros

Restrict the sysconfig renderer availabily to known distros.
Ubuntu/Debian systems may include network-manager but they do
not have support for reading sysconfig network output; that is
enabled via a Network-Manager plugin: ifcfg-rh which is not
available in Ubuntu/Debian.

LP: #1819994

chadmiller-amzn committed:
tools/read-version: handle errors

When the cloned branch was not the canonical upstream and tags were not
available, tox would fail because tools/read-version would fail, and
tragically never print the advice that is in tools/read-version about
how to fix it.

This changes tools/read-version to catch the exception that is elsewhere
explicitly thrown and treat that too as an error it can handle.

smoser committed:
git tests: no longer show warning about safe yaml.

Currently on 18.04, running tox -e py27 will spew errors like:
.tests/unittests/test_net.py:2649: YAMLLoadWarning: calling yaml.load()
without Loader=... is deprecated, as the default Loader is unsafe.
Please read https://msg.pyyaml.org/load for full details.

The change here just uses cloud-init's yaml, which does safeloading
by default.

anhvoms committed:
Azure: Changes to the Hyper-V KVP Reporter

 + Truncate KVP Pool file to prevent stale entries from
   being processed by the Hyper-V KVP reporter.
 + Drop filtering of KVPs as it is no longer needed.
 + Batch appending of existing KVP entries.

anhvoms committed:
DataSourceAzure: Adjust timeout for polling IMDS

If the IMDS primary server is not available, falling back to the
secondary server takes about 1s. The net result is that the
expected E2E time is slightly more than 1s. This change increases
the timeout to 2s to prevent the infinite loop of timeouts.

Commit:
replace remaining occurrences of LOG.warn

jasonzio committed:
cc_mounts: check if mount -a on no-change fstab path

Under some circumstances, cc_disk_setup may reformat volumes which
already appear in /etc/fstab (e.g. Azure ephemeral drive is reformatted
from NTFS to ext4 after service-heal). Normally, cc_mounts only calls
mount -a if it altered /etc/fstab. With this change cc_mounts will read
/proc/mounts and verify if configured mounts are already mounted and if
not raise flag to request a mount -a.  This handles the case where no
changes to fstab occur but a mount -a is required due to change in
underlying device which prevented the .mount unit from running until
after disk was reformatted.

LP: #1825596

Commit:
tests: add Eoan release

goneri committed:
freebsd: add chpasswd pkg in the image

cc_set_passwords.py depends on chpasswd binary.

blackboxsw committed:
release 19.1

Bump the version on cloudinit/version.py to be 19.1 and update ChangeLog

LP: #1828479

jasonzio committed:
Azure: Return static fallback address as if failed to find endpoint

The Azure data source helper attempts to use information in the dhcp
lease to find the Wireserver endpoint (IP address). Under some unusual
circumstances, those attempts will fail. This change uses a static
address, known to be always correct in the Azure public and sovereign
clouds, when the helper fails to locate a valid dhcp lease. This
address is not guaranteed to be correct in Azure Stack environments;
it's still best to use the information from the lease whenever possible.

goneri committed:
freebsd: NoCloud data source support

blkid is a Linux-only command. With this patch, cloud-init uses another
approach to find the data source on FreeBSD.

LP: #1645824

goneri committed:
freebsd: ability to grow root file system

- UFS file system support
- GPT partition table support
- add support for newfs's -L parameter (label)
- move freebsd specific test from Azure to freebsd

goneri committed:
freebsd: fix the name of cloudcfg VARIANT

config/cloud.cfg.tmpl uses 'freebsd', not 'bsd' to identify FreeBSD.

raharper committed:
netplan: update netplan key mappings for gratuitous-arp

Previous versions of netplan included a misspelling for the
bond parameter around gratuitous-arp.  This has been fixed and released
and cloud-init needs to accept both values.  This branch fixes the
key that will be rendered and transforms the previous misspelling
when capturing network_state.

LP: #1827238

Commit:
run-container: centos: comment out the repo mirrorlist

In this way only the 'baseurl' mirror is used, which is easier
to allow through firewalls and proxies.

murraybd committed:
Fix spelling error making 'an Ubuntu' consistent.

markvoelker committed:
Allow identification of OpenStack by Asset Tag

When OpenStack is deployed on some hypervisors (such as VMware
vSphere), cloud-init doesn't detect that it needs to probe the
metadata service because the DMI product name field can't be set
to a field that is recognized by cloud-init.  However, the asset
tag field can be set via flavor extra specs or image metadata.
A similar approach is already used to identify Open Telekom Cloud.

This patch allows cloud init to recognize "OpenStack Nova" or
"OpenStack Compute" in the asset tag field as an indication that
the instance being configured is running on an OpenStack platform.

LP: #1669875

goneri committed:
tools/build-on-freebsd: update to python3

- use python3 by default
- ability to use any Python version through the PYTHON env-var
- indent with 4 spaces
- use 'set -eux'
- remove trailing whitespace
- drop the cheetah dep, Jinja2 is enough

falzm commented:
Sorry I fudged this branch by trying to sync with upstream... 😞

Anyway, the [upstream integration process](https://cloudinit.readthedocs.io/en/latest/topics/hacking.html) goes through Launchpad, so we can close this PR now.
